File: translate.js
import * as deepl from 'deepl-node';
import * as fs from 'fs/promises';

// --- CONFIGURATION ---
// ⚠️ IMPORTANT: Replace with your actual DeepL API Key
const AUTH_KEY = '4d3c30c3-4978-45ed-9bde-a49933fb115b:fx'; 

// The local file to translate
const SOURCE_FILE = 'en.json';

// Define the target languages (DeepL language codes)
const TARGET_LANGUAGES = [
    { code: 'de', name: 'German' },
    { code: 'ja', name: 'Japanese' },
    { code: 'fr', name: 'French' },
    { code: 'ar', name: 'Arabic' }
];
// ---------------------

// Flatten a nested object into a single level with dot-separated keys
function flattenObject(obj, prefix = '') {
    return Object.keys(obj).reduce((acc, k) => {
        const pre = prefix.length ? prefix + '.' : '';
        if (typeof obj[k] === 'object' && obj[k] !== null && !Array.isArray(obj[k])) {
            Object.assign(acc, flattenObject(obj[k], pre + k));
        } else {
            acc[pre + k] = obj[k];
        }
        return acc;
    }, {});
}

// Unflatten an object with dot-separated keys back into a nested object
function unflattenObject(obj) {
    const result = {};
    for (const i in obj) {
        const keys = i.split('.');
        keys.reduce((acc, key, j) => {
            if (j === keys.length - 1) {
                acc[key] = obj[i];
            } else {
                acc[key] = acc[key] || {};
            }
            return acc[key];
        }, result);
    }
    return result;
}

async function translateJsonFile() {
    // 1. Initialize DeepL Translator
    try {
        const translator = new deepl.Translator(AUTH_KEY);
        console.log('DeepL Translator initialized.');
        
        // 2. Load the source JSON data
        console.log(`Reading source file: ${SOURCE_FILE}`);
        const fileContent = await fs.readFile(SOURCE_FILE, 'utf-8');
        const sourceData = JSON.parse(fileContent);

        // 3. Iterate through target languages and translate
        for (const lang of TARGET_LANGUAGES) {
            console.log(`
--- Translating to ${lang.name} (${lang.code.toUpperCase()}) ---`);
            
            const flatData = flattenObject(sourceData);
            
            const keysToTranslate = Object.keys(flatData).filter(key => 
                typeof flatData[key] === 'string' && flatData[key].trim().length > 0
            );

            const sourceTexts = keysToTranslate.map(key => flatData[key]);
            
            if (sourceTexts.length === 0) {
                console.log('No valid strings found to translate. Skipping.');
                continue;
            }

            try {
                const results = await translator.translateText(
                    sourceTexts, 
                    'en', 
                    lang.code,
                    { 
                        tagHandling: 'xml',
                        formality: (lang.code === 'ar') ? 'prefer_more' : undefined 
                    }
                );
                
                const translatedFlatData = { ...flatData };
                for (let i = 0; i < keysToTranslate.length; i++) {
                    translatedFlatData[keysToTranslate[i]] = results[i].text;
                }
                
                const translatedData = unflattenObject(translatedFlatData);

                // 4. Save the translated JSON file
                const outputFileName = `${lang.code.toLowerCase()}.json`;
                await fs.writeFile(
                    outputFileName, 
                    JSON.stringify(translatedData, null, 4), 
                    'utf-8'
                );
                console.log(`✅ Successfully saved translation to ${outputFileName}`);

            } catch (error) {
                console.error(`❌ DeepL Error during translation to ${lang.name}: ${error.message}`);
            }
        }
    } catch (error) {
        console.error(`❌ Fatal Error in script execution: ${error.message}`);
    }
}

translateJsonFile();

File: translate.py
import deepl
import json
import os

# --- CONFIGURATION ---
# Replace 'YOUR_DEEPL_API_KEY' with your actual DeepL API key
API_KEY = "4d3c30c3-4978-45ed-9bde-a49933fb115b:fx"

# The path to your source English JSON file
SOURCE_FILE = "en.json"

# List of target languages (DeepL language codes)
# The output file will be named based on the code (e.g., 'de.json')
TARGET_LANGUAGES = {
    "DE": "German",
    "JA": "Japanese",
    "FR": "French",
    "AR": "Arabic"
}
# ---------------------

def translate_json_file():
    """Reads the source JSON, translates its values, and saves new language JSONs."""
    
    # 1. Initialize DeepL Translator
    try:
        translator = deepl.Translator(API_KEY)
        # Verify the key (optional, but good practice)
        usage = translator.get_usage()
        if usage.any_limit_reached:
            print(f"⚠️ DeepL usage limit reached. Cannot proceed.")
            return
        print(f"DeepL API Key is valid. Characters remaining this month: {usage.character.limit - usage.character.count}")
    except Exception as e:
        print(f"❌ Error initializing DeepL translator: {e}")
        print("Please check your API key.")
        return

    # 2. Load the source JSON file
    try:
        with open(SOURCE_FILE, 'r', encoding='utf-8') as f:
            source_data = json.load(f)
        print(f"Successfully loaded source file: {SOURCE_FILE}")
    except FileNotFoundError:
        print(f"❌ Error: {SOURCE_FILE} not found. Make sure the file is in the same directory.")
        return
    except json.JSONDecodeError:
        print(f"❌ Error: Could not decode JSON from {SOURCE_FILE}. Check the file format.")
        return
    
    # 3. Iterate through target languages and translate
    for target_code, lang_name in TARGET_LANGUAGES.items():
        print(f"\n--- Translating to {lang_name} ({target_code}) ---")
        
        translated_data = {}
        
        # Iterate over key-value pairs in the source JSON
        for key, value in source_data.items():
            if isinstance(value, dict):
                translated_data[key] = {}
                for sub_key, sub_value in value.items():
                    if isinstance(sub_value, dict):
                        translated_data[key][sub_key] = {}
                        for sub_sub_key, sub_sub_value in sub_value.items():
                            if isinstance(sub_sub_value, str) and sub_sub_value.strip():
                                try:
                                    # DeepL translation call
                                    result = translator.translate_text(
                                        sub_sub_value,
                                        source_lang="EN",
                                        target_lang=target_code,
                                        tag_handling="xml" # Good practice for translation of HTML/XML-like content
                                    )
                                    translated_data[key][sub_key][sub_sub_key] = result.text
                                except deepl.DeepLError as e:
                                    print(f"⚠️ DeepL Error for key '{key}.{sub_key}.{sub_sub_key}': {e}. Skipping this translation.")
                                    translated_data[key][sub_key][sub_sub_key] = sub_sub_value # Fallback to original text
                                except Exception as e:
                                    print(f"❌ An unexpected error occurred for key '{key}.{sub_key}.{sub_sub_key}': {e}. Skipping.")
                                    translated_data[key][sub_key][sub_sub_key] = sub_sub_value
                            else:
                                translated_data[key][sub_key][sub_sub_key] = sub_sub_value
                    elif isinstance(sub_value, str) and sub_value.strip():
                        try:
                            # DeepL translation call
                            result = translator.translate_text(
                                sub_value,
                                source_lang="EN",
                                target_lang=target_code,
                                tag_handling="xml" # Good practice for translation of HTML/XML-like content
                            )
                            translated_data[key][sub_key] = result.text
                        except deepl.DeepLError as e:
                            print(f"⚠️ DeepL Error for key '{key}.{sub_key}': {e}. Skipping this translation.")
                            translated_data[key][sub_key] = sub_value # Fallback to original text
                        except Exception as e:
                            print(f"❌ An unexpected error occurred for key '{key}.{sub_key}': {e}. Skipping.")
                            translated_data[key][sub_key] = sub_value
                    else:
                        translated_data[key][sub_key] = sub_value
            elif isinstance(value, str) and value.strip():
                try:
                    # DeepL translation call
                    result = translator.translate_text(
                        value,
                        source_lang="EN",
                        target_lang=target_code,
                        tag_handling="xml" # Good practice for translation of HTML/XML-like content
                    )
                    translated_data[key] = result.text
                except deepl.DeepLError as e:
                    print(f"⚠️ DeepL Error for key '{key}': {e}. Skipping this translation.")
                    translated_data[key] = value # Fallback to original text
                except Exception as e:
                    print(f"❌ An unexpected error occurred for key '{key}': {e}. Skipping.")
                    translated_data[key] = value
            else:
                # Handle non-string or empty values (e.g., nested objects, numbers)
                translated_data[key] = value

        # 4. Save the translated JSON file
        output_file = f"{target_code.lower()}.json"
        try:
            with open(output_file, 'w', encoding='utf-8') as f:
                json.dump(translated_data, f, indent=4, ensure_ascii=False)
            print(f"✅ Successfully saved translation to {output_file}")
        except Exception as e:
            print(f"❌ Error saving file {output_file}: {e}")


if __name__ == "__main__":
    translate_json_file()